shapely/_flatcoords.py

```diff
@@ -0,0 +1,426 @@
+        geometry type. See
+        https://github.com/geoarrow/geoarrow/blob/main/format.md for details.
+
+    See also
+    --------
+    from_ragged_array
+
+    Examples
+    --------
+    Consider a Polygon with one hole (interior ring):
+
+    >>> import shapely
+    >>> polygon = shapely.Polygon(
+    ...     [(0, 0), (10, 0), (10, 10), (0, 10)],
+    ...     holes=[[(2, 2), (3, 2), (2, 3)]]
+    ... )
+    >>> polygon
+    <POLYGON ((0 0, 10 0, 10 10, 0 10, 0 0), (2 2, 3 2, 2 3, 2 2))>
+
+    This polygon can be thought of as a list of rings (first ring the exterior
+    ring, subsequent rings the interior rings), and each ring as a list of
+    coordinate pairs. This is very similar to how GeoJSON represents the
+    coordinates:
+
+    >>> import json
+    >>> json.loads(shapely.to_geojson(polygon))["coordinates"]
+    [[[0.0, 0.0], [10.0, 0.0], [10.0, 10.0], [0.0, 10.0], [0.0, 0.0]],
+     [[2.0, 2.0], [3.0, 2.0], [2.0, 3.0], [2.0, 2.0]]]
+
+    This function will semantically return such a list of lists of lists, but
+    using a single contiguous array of coordinates, and multiple arrays of
+    offsets:
+
+    >>> typ, coords, offsets = shapely.to_ragged_array([polygon])
+    >>> typ
+    <GeometryType.POLYGON: 3>
+    >>> coords
+    array([[ 0.,  0.],
+           [10.,  0.],
+           [10., 10.],
+           [ 0., 10.],
+           [ 0.,  0.],
+           [ 2.,  2.],
+           [ 3.,  2.],
+           [ 2.,  3.],
+           [ 2.,  2.]])
+
+    >>> offsets
+    (array([0, 5, 9]), array([0, 2]))
+
+    As an example how to interpret the offsets: the i-th ring in the
+    coordinates is represented by ``offsets[0][i]`` to ``offsets[0][i+1]``:
+
+    >>> ring1_start, ring1_end = offsets[0][0], offsets[0][1]
+    >>> coords[ring1_start:ring1_end]
+    array([[ 0.,  0.],
+           [10.,  0.],
+           [10., 10.],
+           [ 0., 10.],
+           [ 0.,  0.]])
+
+    """
+    geom_types = np.unique(get_type_id(arr))
+    # ignore missing values (type of -1)
+    geom_types = geom_types[geom_types >= 0]
+
+    if len(geom_types) == 1:
+        typ = GeometryType(geom_types[0])
+        if typ == GeometryType.POINT:
+            coords, offsets = _get_arrays_point(arr, include_z)
+        elif typ == GeometryType.LINESTRING:
+            coords, offsets = _get_arrays_linestring(arr, include_z)
+        elif typ == GeometryType.POLYGON:
+            coords, offsets = _get_arrays_polygon(arr, include_z)
+        elif typ == GeometryType.MULTIPOINT:
+            coords, offsets = _get_arrays_multipoint(arr, include_z)
+        elif typ == GeometryType.MULTILINESTRING:
+            coords, offsets = _get_arrays_multilinestring(arr, include_z)
+        elif typ == GeometryType.MULTIPOLYGON:
+            coords, offsets = _get_arrays_multipolygon(arr, include_z)
+        else:
+            raise ValueError(f"Geometry type {typ.name} is not supported")
+
+    elif len(geom_types) == 2:
```

Comment:
Is there a specific use case to be that tolerant here? In general I would say no to these kind of exceptions.

If there is a use case, then I'd prefer having a separate "force_collection" function to make this conversion explicit (instead of silently converting here)

Reply:
I think a very concrete use case is that many shapely functions that create geometries will actually give you a mixture (GEOS is typically not strict about the type it returns). For example if you are overlaying polygons, you will almost always get a mix of Polygons and MultiPolygons. So my experience is that it is very normal to have such a mixture. 

The result would also be exactly the same, so there is no ambiguity in doing it (except that users might unexpectedly get multipolygon-formatted results if they thought to only have polygons).
(it also follows naturally from other functions we have like `get_num_geometries` which also work with mixtures)

I would also like that tools that use this function can enable such automatic behaviour (eg geopandas using this to convert geometries to arrow to then write to file). Of course, they could make use of a `force_collection` function, but that does require another pass over the data and cloning of geometries. While if it is integrated in this function, it can be done in one pass.

So I would personally like to have at least the option to do it within `to_ragged_array` (to be able to it as efficiently as possible). But it could be behind a flag? (in pyogrio we have a `promote_to_multi` keyword, which is True by default (if the file format requires it))

Reply:
And BTW, we should certainly have a "force_collection" like function, that would be handy, regardless of the above discussion (PostGIS has this as https://postgis.net/docs/ST_Multi.html)

Reply:
I am :+1: on making this an option (and also on defaulting this to `True`)

I am mildly concerned that this doesn't roundtrip and that the outcome of a roundtrip of for instance a Polygon will depend on other elements in the array being a Multi- or not.

Reply:
> I am mildly concerned that this doesn't roundtrip and that the outcome of a roundtrip of for instance a Polygon will depend on other elements in the array being a Multi- or not.

Yes, rountripping a mixture of singular/multi geoms is indeed not exact. I asked about this in the top post:

> For roundtripping mixed single and multi geometries, would we like to have an option in the `from_` function to return Multi geometries with 1 part as the singular version? (that doesn't guarantee a faithful rountrip, since you can have eg MultiPolygons with a single sub-polygon in your input, but I do think it would cover the large majority of use cases where someone would prefer to preserve the mixture).

We also had a long discussion about this in pyogrio: some GDAL-supported file formats requires a single geometry type, and so we convert singular/multi of the same type when writing, but with the consequence that reading back doesn't exactly preserve the types. I think among others Levi argues that this isn't typically that important (as long as it are always single/multi of the same type)

Reply:
Ok 👍

Reply:
If we add a `promote_to_multi` keyword, then that gives also some more complication / questions: what does `promote_to_multi=True` mean if you have only Polygons? Also output as MultiPolygon? 
And if so, we would have a default of `promote_to_multi=None` that means "only promote if you actually have a mixture", allowing the user to override with True or False to always be sure about what happens with singular types?

Reply:
> And if so, we would have a default of promote_to_multi=None that means "only promote if you actually have a mixture", allowing the user to override with True or False to always be sure about what happens with singular types?

Yes I think that is precisely what should happen; an "automatic promotion if necessary" by default (like you implemented in this PR) and if it is True/False, make sure the output is always a collection (or not).

But to be clear , that effort can be separate from this PR, because it can be done in a backwards compatible way.

Reply:
> But to be clear , that effort can be separate from this PR, because it can be done in a backwards compatible way.

Yes, that's right. If we have the default to autopromote if necessary/possible (current behaviour), then this can be done as an addition later on.

I opened https://github.com/shapely/shapely/issues/1617 to discuss the "force_collection" idea.